text/0000-inline-buildpack.md

```diff
@@ -0,0 +1,126 @@
+# Meta
+[meta]: #meta
+- Name: Inline Buildpacks
+- Start Date: 2020-06-11
+- Author(s): [Joe Kutner](https://github.com/jkutner)
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: N/A
+
+# Summary
+[summary]: #summary
+
+This is a proposal for a type of buildpack that can be defined in the same repo as the app it is used with.
+
+# Motivation
+[motivation]: #motivation
+
+One characteristic of `Dockerfile` that Buildpacks do not replace well is the ability to create a single, one-off file to quickly define a custom build process for an app or project. A `Dockerfile` may contain just a few lines of code, and a developer can generate a complete image. The only quick way to create a custom build for an app using a buildpack is by creating no fewer than three files: `buildpack.toml`, `bin/build`, and `bin/detect` (and likely storing those in a new repo).
+
+We seek to reduce the barrier to entry for buildpacks when a developer needs to quickly create a custom, one-off build step by introducing a new mechanism that will allow a buildpack to be defined inline with an app (and with fewer files).
+
+# What it is
+[what-it-is]: #what-it-is
+
+- **inline** - existing in the same repository
+
+The target personas for this RFC are the buildpack author and buildpack user who need to create a custom, one-off build or are just getting started with buildpacks and do not want the overhead of a new repo.
+
+We propose two new keys in the `[[build.buildpacks]]` table in `project.toml`:
+
+- `api` - the API compatibility version of the buildpack. Matches the schema of `buildpack.toml`
+- `inline` - the contents of a script that will be run as the `bin/build` of the inline buildpack
+- `shell` - defines the shebang line of the script in `inline`
+
+When any of these keys are provided the following keys are disallowed: `version`, `uri`.
+
+(Note: there is no need to implement `bin/detect`. The detect phase will always pass because there's no foreseeable reason that a user would create an inline buildpack that they don't want to use)
+
+# How it Works
+[how-it-works]: #how-it-works
+
+When an entry in the `[[build.buildpacks]]` table contains an `inline` value, the platform WILL NOT attempt to download the buildpack `id@version`. Instead, it will create an ephemeral buildpack, with the following:
```

Comment:
Could we leave this unspecified? It may be significantly more efficient to have the lifecycle execute `inline` directly with `shell`. This will also make it easier to `CNB_BUILDPACK_DIR` to the app dir, which wouldn't otherwise be possible.

Reply:
Yea, that makes sense and it would resolve the discussion in https://github.com/buildpacks/rfcs/pull/86#discussion_r450251899. But then we're talking about lifecycle understanding this inline construct.

Reply:
If we specify an `inline` version of `buildpack.toml` we could solve both of these problems? The inline buildpack in `project.toml` could be tranforms into an "inline" `buildpack.toml`. https://github.com/buildpacks/rfcs/pull/86#discussion_r451643134

Reply:
@ekcasey i want `inline` in `buildpack.toml` too, but was hoping to punt on that for now. I think i can leave this unspecified, and we can deal with it later.

Reply:
addressed in de7783a